### PR TITLE
Initial public draft for TotW 131.

### PR DESCRIPTION
**Pull request by tituswinters**

### CODE REVIEW

File: _posts/2017-12-08-totw-131.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+---
+title: "Tip of the Week #131: Special Member Functions and `= default`"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/131
+type: markdown
+---
```
Comment (manshreck) [inline _posts/2017-12-08-totw-131.md]: Add order: "131"

File: _posts/2017-12-08-totw-131.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+give control over when those defaults are declared and defined.
+
+## What Does `=default` Do, and Why Would We Use It?
+
+Writing `=default` is our way to tell the compiler "do what you would normally
+have done for this special member function". Why might we want to do this
+rather than writing an implementation by hand or leaving the compiler to
+declare one for us?
+
+*   We can change the access level (e.g., make a constructor `protected` instead
```
Comment (manshreck) [inline _posts/2017-12-08-totw-131.md]: Nit: don't think the bullet points actually warrant semi-colon endings. -> periods.

File: _posts/2017-12-08-totw-131.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+*   Compiler-provided special member functions can be *trivial* (when all of the
+    operations they invoke are themselves trivial), which can make them faster
+    and safer;
+*   Types with defaulted constructors can be *aggregates*, and hence support
+    *aggregate initialization*, whereas those with user-provided constructors
+    cannot.
+*   Explicitly declaring a defaulted member gives us a place to document the
+    semantics of the resulting function;
+*   In a class template, `=default` gives a simple way to conditionally provide
+    an operation depending on whether some underlying type provides it.
```
Comment (manshreck) [inline _posts/2017-12-08-totw-131.md]: Awkward sentence. One use of "give" and two "provides" ...

How about:

"In a class template, `=default` is a simple way to declare an operation
conditionally, based on whether some underlying type provides it."

File: _posts/2017-12-08-totw-131.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+ public:
+  D() = default;  // unsurprising, but not explicit: D has a default constructor
+ private:
+  std::unique_ptr<int> p;  // std::unique_ptr has a default constructor
+};
+```
+
+Clearly we shouldn't write code like `class C`: in a non-template, use
+`=default` only if you intend the class to support the operation (and then test
+that it does).  clang-tidy includes a check for this.
```
Comment (manshreck) [inline _posts/2017-12-08-totw-131.md]: bold and code clang-tidy? (That's what I usually do for tools/command-lines.)

*`clang-tidy*`

File: _posts/2017-12-08-totw-131.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+When `=default` is used *after* the first declaration of a special member
+function (i.e., outside of the class), it has a simpler meaning: it tells the
+compiler to define the function, and to give an error if it is unable to do so.
+When `=default` is used outside of the class, the defaulted function will not be
+trivial: triviality is determined by the first declaration (so that all clients
+agree on whether the operation is trivial or not).
+
+If you don't need your class to be an aggregate and you don't need the
+constructor to be trivial then defaulting the constructor outside of the class
+definition like examples `E` and `F` below is often a good choice.  Its meaning
```
Comment (manshreck) [inline _posts/2017-12-08-totw-131.md]: , like in examples `E` and `F` below,

(add commas to clarify the sentence flow.)

File: _posts/2017-12-08-totw-131.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+```
+
+From C++11 onwards we have more options.
+
+```c++
+class C {
+ public:
+  C() = default;  // misleading: C has a deleted default constructor
+ private:
+  const int i;  // const => must always be initialized.
```
Comment (manshreck) [inline _posts/2017-12-08-totw-131.md]: Nit: when possible, I try to align comments with comments in the same code block. Not important here. More readable, IMO, in the code blocks below.

File: _posts/2017-12-08-totw-131.md (hunk 3 above)
Comment (tituswinters) [reply _posts/2017-12-08-totw-131.md]: Done.

File: _posts/2017-12-08-totw-131.md (hunk 2 above)
Comment (tituswinters) [reply _posts/2017-12-08-totw-131.md]: Done.

File: _posts/2017-12-08-totw-131.md (hunk 6 above)
Comment (tituswinters) [reply _posts/2017-12-08-totw-131.md]: this is what clang-format would produce.

File: _posts/2017-12-08-totw-131.md (hunk 4 above)
Comment (tituswinters) [reply _posts/2017-12-08-totw-131.md]: Done.

File: _posts/2017-12-08-totw-131.md (hunk 5 above)
Comment (tituswinters) [reply _posts/2017-12-08-totw-131.md]: Done.

File: _posts/2017-12-08-totw-131.md (hunk 1 above)
Comment (tituswinters) [reply _posts/2017-12-08-totw-131.md]: Done.

**Comment by tituswinters** [comment]
PTAL.

**Pull request merged**